### PR TITLE
feat(ec2): add d3 and d3en instances

### PR DESCRIPTION
**Pull request by robertd**
`d3` and `d3en` storage optimized instances:
https://aws.amazon.com/blogs/aws/ec2-update-d3-d3en-dense-storage-instances/

image

image

CFN docs have already been updated:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#cfn-ec2-instance-instancetype

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by robertd** [comment]
Related to: https://github.com/aws/jsii/issues/3208

**Comment by robertd** [comment]
https://github.com/aws/jsii/pull/3209 landed... so this will probably go away once jsii gets bumped to `1.47.0` whenever that one gets released.

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 1]
```diff
@@ -332,6 +332,21 @@ export enum InstanceClass {
...
+
+  /**
+   * Storage-optimized instances, 3rd generation
+   */
+  D3 = 'd3',
+
+  /**
+  * Storage-optimized instances, 3rd generation
+  */
+  D3EN = 'd3en',
```
Comment (jumic) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:348]: Could you please add a symbolic name for this enum, too? Thanks.

Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:348]: Can you suggest one?

Comment (jumic) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:348]: The difference between the two instance types is described on the [product page](https://aws.amazon.com/de/ec2/instance-types/d3/): 
> D3en instances offer a 3.5x higher storage-to-vCPU ratio compared to D2 instances.

Maybe we could use `STORAGE3_ENHANCED_STORAGE` or just `STORAGE3_ENHANCED`?

Or does someone else have a suggestion?

**Comment by robertd** [comment]
Depends on #17904

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts (hunk 1 above)
Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:358]: @njlynch wdyt?

Comment (njlynch) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:358]: I'm a bit torn on having the longer/symbolic names, given that we effectively just make up the names for the suffixes, and there's no consistency or guidelines. That being said, we've started down this road, so we might as well continue for the time being.

It's hard to pin down what `en` stands for from the blog post, but I _think_ it stands for "elastic networking".

> The D3en instances have a high ratio of storage to vCPU, and are optimized for high throughput and high sequential I/O to very large data sets, with a cost-per-TB that is 80% lower than on D2 instances.
> AMIs – You must use AMIs that include the Elastic Network Adapter (ENA) and NVMe drivers.

That's my guess from the above, at least. That would imply `STORAGE3_ELASTIC_NETWORK` would be the name.

Comment (njlynch) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:358]: Ah, this seems to be the official documentation on the suffixes: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#instance-hardware-specs

So:
* `e – Extra storage or memory`
* `n – Network optimization`

`STORAGE3_ENHANCED_NETWORK` might be a decent pick. Or `STORAGE3_EXTRA_STORAGE_NETWORK`.

Comment (robertd) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:363]: Personally I never use symbolic names :D. Thanks for the input @njlynch. PR has been updated.

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 2]
```diff
@@ -342,6 +342,26 @@ export enum InstanceClass {
...
+
+  /**
+   * Storage-optimized instances, 3rd generation
+   */
+  D3 = 'd3',
+
+  /**
+  * Storage-optimized instances, 3rd generation
+  */
+  STORAGE3_ELASTIC_NETWORK = 'd3en',
```
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:358]:
```suggestion
  STORAGE3_ENHANCED_NETWORK = 'd3en',
```

**Review by njlynch — APPROVED** [review]

**Pull request merged**